### PR TITLE
Add optimizer rules for multiplication and division by left-shifted one.

### PR DESCRIPTION
**Pull request by ekpyron**
@leonardoalt I would expect these to be easy to verify by SMT?

### CODE REVIEW

**Comment by leonardoalt** [comment]
Not sure, involves nonlinear arithmetic. But I'll write the code

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -391,6 +391,31 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		false
 	});
 
+	rules.push_back({
+		// MUL(X, SHL(1, Y)) -> SHL(X, Y)
```
Comment (chriseth) [inline libevmasm/RuleList.h]: Arguments for shift need to be reversed.

Comment (ekpyron) [reply libevmasm/RuleList.h]: Haha, duh, right, yes!

**Comment by leonardoalt** [comment]
https://github.com/leonardoalt/solidity_proofs/pull/2

The mul rules were proved for 64bit and div for 32.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**